Source/bmalloc/libpas/src/libpas/pas_mar_registry.c

```diff
@@ -0,0 +1,233 @@
+    if (hash == registry->backtrace_registry[index].hash)
+        return index;
+
+    registry->backtrace_registry[index].num_frames = num_stack_frames;
+    registry->backtrace_registry[index].hash = hash;
+    memcpy(registry->backtrace_registry[index].backtrace_buffer, backtrace, num_stack_frames* sizeof(void*));
+    return index;
+}
+
+// MAR Registry
+
+void pas_mar_initialize(void)
+{
+    if (!pas_mar_registry_for_crash_reporter_enumeration)
+        pas_mar_registry_for_crash_reporter_enumeration = &pas_mar_global_registry;
+
+    if (PAS_UNLIKELY(getenv("SanitizersAllocationTraces"))) {
+        pas_mar_enabled = true;
+        return;
+    }
+    if (PAS_LIKELY(pas_get_fast_random(1000) >= 1))
+        pas_mar_enabled = false;
+    else {
+        pas_mar_enabled = true;
+        pas_mar_qualifying_page_index = pas_get_fast_random(PAS_MAR_PROBABILITY);
+    }
+}
+
+void* pas_mar_did_allocate(pas_mar_registry* registry, void* address, size_t allocation_size)
+{
+    void* stacktrace[PAS_MAR_BACKTRACE_MAX_SIZE];
+    unsigned num_stack_frames = (unsigned) backtrace(stacktrace, PAS_MAR_BACKTRACE_MAX_SIZE);
+
+    return pas_mar_record_allocation(registry, address, allocation_size, num_stack_frames, stacktrace);
+}
+
+void* pas_mar_did_allocate_and_zero(pas_mar_registry* registry, pas_allocation_result result, size_t allocation_size)
+{
+    void* stacktrace[PAS_MAR_BACKTRACE_MAX_SIZE];
+    unsigned num_stack_frames = (unsigned) backtrace(stacktrace, PAS_MAR_BACKTRACE_MAX_SIZE);
+
+    pas_mar_record_allocation(registry, (void*)result.begin, allocation_size, num_stack_frames, stacktrace);
+    return (void*)pas_allocation_result_zero(result, allocation_size).begin;
+}
+
+void* pas_mar_did_deallocate(pas_mar_registry* registry, void* address)
+{
+    void* stacktrace[PAS_MAR_BACKTRACE_MAX_SIZE];
+    unsigned num_stack_frames = (unsigned) backtrace(stacktrace, PAS_MAR_BACKTRACE_MAX_SIZE);
+
+    return pas_mar_record_deallocation(registry, address, num_stack_frames, stacktrace);
+}
+
+void* pas_mar_record_allocation(pas_mar_registry* registry, void* address, size_t allocation_size, unsigned num_stack_frames, void** backtrace)
+{
+    PAS_ASSERT(num_stack_frames <= PAS_MAR_BACKTRACE_MAX_SIZE);
+
+    if (pas_mar_is_allocation_table_full(registry))
+        pas_mar_increment_allocation_record_table_head(registry);
+
+    unsigned allocation_table_index = pas_mar_allocation_table_tail_index(registry);
+    pas_mar_increment_allocation_record_table_tail(registry);
+
+    uint32_t backtrace_hash = hash_backtrace(num_stack_frames, backtrace);
+    unsigned backtrace_registry_index = pas_mar_insert_backtrace(registry, num_stack_frames, backtrace, backtrace_hash);
+    struct pas_mar_memory_action_record new_record = {
+        address,
+        allocation_size,
+        backtrace_registry_index,
+        backtrace_hash,
+        true
+    };
+    registry->allocation_record_table[allocation_table_index] = new_record;
```

Comment:
This seems like it needs release semantics -- otherwise another thread could read a table index and only see some of the previous writes, thereby getting a half-valid or torn pas_mar_memory_action_record.

Reply:
Correspondingly anywhere that reads from the allocation_record_table needs to do an atomic acquire